### PR TITLE
Adding TorchEmbedding contrib op

### PR DESCRIPTION
**Pull request by neginraoof**
**Description**: Adding TorchEmbedding contrib op based on pytorch embeddin operator.

**Motivation and Context**
- Why is this change required? What problem does it solve?
This pytorch operator has a similar forward logic as Gather. However, there's a difference in the behavior of the op for updating gradients in the backward pass. The two input arguments 'padding_idx' and 'scale_grad_by_freq' are arguments that control the behavior of this op for updating grad, and hence we need a specific operator in ONNX which allows us to control backward logic.

### CODE REVIEW

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 1]
```diff
@@ -2290,6 +2290,83 @@ It's an extension of Gelu. It takes the sum of input A and bias input B as the i
...
+          "i.e. it remains as a fixed pad."
+          "tensor(int64)",
+          OpSchema::Optional)
+      .Input(
+          3,
+          "scale_grad_by_freq",
+          "A 0-D bool tensor. If given, this will scale gradients by the inverse of frequency of "
+          "the indices (words) in the mini-batch. Default  is ``False``",
+          "tensor(bool)",
+          OpSchema::Optional)
```
Comment (ashbhandare) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2327]: Comparing with torch.nn.functional.embedding, we are missing a few inputs, particularly max_norm and norm_type. Will the exporter take care of adding the appropriate nodes downstream to this op for normalization if these are provided?

File: onnxruntime/test/contrib_ops/shape_inference_test_helper.h [hunk 2]
```diff
@@ -78,6 +79,7 @@ inline void TestShapeInference(
   for (auto const& n_ : inputs) {
     node.add_input(n_.name());
     *graph->add_input() = n_;
+    std::cout << "nnn  " << n_.name() << "\n";
```
Comment (ashbhandare) [inline onnxruntime/test/contrib_ops/shape_inference_test_helper.h:82]: unintentional?

Comment (neginraoof) [reply onnxruntime/test/contrib_ops/shape_inference_test_helper.h]: Thanks!

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 1 above)
Comment (neginraoof) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2327]: These arguments are present in at python level, but the backend op does not receive these arguments.
Here is the ATen op implementation: https://github.com/pytorch/pytorch/blob/a28c7db9f906a3dd445094e670b9b80b2feec9ec/aten/src/ATen/native/Embedding.cpp#L16

So, when converting the python module, the corresponding ops will be added.

Comment (mrry) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2327]: Thanks for clarifying this, Negin! Your reply prompted me to look into the implementation of `max_norm`:

https://github.com/pytorch/pytorch/blob/645119eaefd0dcf1afc54e9ad58678b5245dea78/torch/nn/functional.py#L1960-L1962

I don't think the right thing will happen if we try to export with `max_norm` set to anything other than `None` - even after adding the contrib op - because the definition relies on modifying the embedding matrix in place, and I don't think we have a way to represent that in the ONNX graph. 

Not a blocker for this PR, but do you think there is any way we could raise an error when we hit this case? (/cc @spandantiwari for extra eyes on the problem).

File: onnxruntime/test/contrib_ops/shape_inference_test_helper.h [hunk 3]
```diff
@@ -91,8 +92,7 @@ inline void TestShapeInference(
   ONNX_NAMESPACE::shape_inference::InferShapes(model, false, schema_registry);
 
   auto inferredGraph = model.graph();
-  int index = static_cast<int>(inputs.size());  // index for value_info of output
-  auto inferred_output = inferredGraph.value_info(index);
+  auto inferred_output = inferredGraph.value_info(1);
```
Comment (ashbhandare) [inline onnxruntime/test/contrib_ops/shape_inference_test_helper.h:95]: is this change correct?

File: onnxruntime/test/contrib_ops/torch_embedding_shape_inference_test.cc [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  ONNX_NAMESPACE::ValueInfoProto input;
+  CreateValueInfo(input, "indices", ONNX_NAMESPACE::TensorProto_DataType_INT64, shape);
+  std::vector<int64_t> shape_w = {10, 3};
+  ONNX_NAMESPACE::ValueInfoProto weight;
+  CreateValueInfo(weight, "weight", ONNX_NAMESPACE::TensorProto_DataType_INT64, shape_w);
+  ONNX_NAMESPACE::ValueInfoProto padding_idx;
+  CreateValueInfo(padding_idx, "Pad", ONNX_NAMESPACE::TensorProto_DataType_INT64, {});
+  ONNX_NAMESPACE::ValueInfoProto scale_grad;
+  CreateValueInfo(scale_grad, "Scale", ONNX_NAMESPACE::TensorProto_DataType_BOOL, {});
+  std::vector<ONNX_NAMESPACE::ValueInfoProto> inputs = {weight, input, padding_idx};
```
Comment (ashbhandare) [inline onnxruntime/test/contrib_ops/torch_embedding_shape_inference_test.cc:97]: missed passing scale_grad input

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 1 above)
Comment (neginraoof) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2327]: So, the op which is used to normalize the Embedding layer is an ATen op called embedding_renorm_.
It is correct that this is an inplace op, but exporter does have some preprocessing to take care of inplace ops.
However, we currently do not have export support for this op. As you mentioned this is outside of the scope of this PR. But we can support for this op in exporter in general. No blockers.

File: onnxruntime/test/contrib_ops/torch_embedding_shape_inference_test.cc (hunk 4 above)
Comment (neginraoof) [reply onnxruntime/test/contrib_ops/torch_embedding_shape_inference_test.cc]: Yes, thanks!

**Comment by neginraoof** [comment]
@ashbhandare @mrry 
I think comments are addressed. Would you please provide your review? Thanks

**Review by mrry — APPROVED** [review]
Looks good to me!

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 1 above)
Comment (mrry) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2327]: Right... I think there might still be a blocker because IIUC the exporter won't export a graph that has a side effect on one of its inputs (i.e. the embedding weights), so the visible side effect of running the exported graph as an ORTModule would be different from running the original module.

**Review by mrry — APPROVED** [review]
